File: frontend/src/types/stats.ts
import type { RankingEntry } from "./ranking";

export interface DashboardStats {
  users_count: number;
  groups_count: number;
  teams_count: number;
  tournaments_count: number;
  matches_count: number;
  scheduled_matches_count: number;
  finished_matches_count: number;
  predictions_count: number;
  total_points_awarded: number;
  average_points_per_prediction: number;
}

export interface GroupStats {
  group_id: number;
  group_name: string;
  members_count: number;
  predictions_count: number;
  total_points: number;
  average_points: number;
  leader_name: string | null;
  leader_points: number | null;
}

export interface StatsOverview {
  dashboard: DashboardStats;
  top_general: RankingEntry[];
  top_exact_scores: RankingEntry[];
  top_winners: RankingEntry[];
  group_stats: GroupStats[];
}

File: backend/app/schemas/stats.py
from pydantic import BaseModel

from app.schemas.ranking import RankingEntryRead


class DashboardStatsRead(BaseModel):
    users_count: int
    groups_count: int
    teams_count: int
    tournaments_count: int
    matches_count: int
    scheduled_matches_count: int
    finished_matches_count: int
    predictions_count: int
    total_points_awarded: int
    average_points_per_prediction: float


class UserStatsRead(BaseModel):
    user_id: int
    full_name: str
    username: str
    total_points: int
    predictions_count: int
    exact_scores_count: int
    winner_count: int
    goal_difference_count: int
    groups_count: int


class GroupStatsRead(BaseModel):
    group_id: int
    group_name: str
    members_count: int
    predictions_count: int
    total_points: int
    average_points: float
    leader_name: str | None
    leader_points: int | None


class StatsOverviewRead(BaseModel):
    dashboard: DashboardStatsRead
    top_general: list[RankingEntryRead]
    top_exact_scores: list[RankingEntryRead]
    top_winners: list[RankingEntryRead]
    group_stats: list[GroupStatsRead]